File: sdks/sandbox/python/src/opensandbox/models/filesystem.py
#
# Copyright 2025 Alibaba Group Holding Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""
Filesystem-related data models.

Models for file operations, directory listings, and filesystem metadata.
"""

from datetime import datetime
from io import IOBase

from pydantic import BaseModel, ConfigDict, Field, field_validator


class EntryInfo(BaseModel):
    """
    Metadata information for a file or directory entry.

    Contains complete filesystem metadata including path, permissions, ownership,
    size, and timestamp information for files and directories in the sandbox.
    """

    path: str = Field(description="Absolute path of the file or directory")
    entry_type: str | None = Field(default=None, description="Entry type", alias="type")
    mode: int = Field(description="Unix file mode/permissions as integer (e.g., 644)")
    owner: str = Field(description="Owner username of the file or directory")
    group: str = Field(description="Group name of the file or directory")
    size: int = Field(description="Size of the file in bytes (0 for directories)")
    modified_at: datetime = Field(
        description="Timestamp when entry was last modified", alias="modified_at"
    )
    created_at: datetime = Field(
        description="Timestamp when entry was created", alias="created_at"
    )

    model_config = ConfigDict(populate_by_name=True)


class WriteEntry(BaseModel):
    """
    Request to write content to a file.

    Creates or overwrites a file with the specified content, permissions, and ownership.
    Supports both text and binary data through flexible data parameter.
    """

    path: str = Field(description="Destination file path where content will be written")
    data: str | bytes | IOBase | None = Field(
        default=None, description="Content to write - can be str or bytes"
    )
    mode: int = Field(default=755, description="Unix file permissions as integer")
    owner: str | None = Field(default=None, description="Owner username to set")
    group: str | None = Field(default=None, description="Group name to set")
    encoding: str = Field(
        default="utf-8", description="Character encoding for string data"
    )
    model_config = ConfigDict(arbitrary_types_allowed=True)

    @field_validator("path")
    @classmethod
    def path_must_not_be_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Path cannot be blank")
        return v

    @field_validator("mode")
    @classmethod
    def mode_must_be_non_negative(cls, v: int) -> int:
        if v < 0:
            raise ValueError("Mode must be non-negative")
        return v

    @field_validator("encoding")
    @classmethod
    def encoding_must_not_be_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Encoding cannot be blank")
        return v


class DirectoryListEntry(BaseModel):
    """Request to list directory contents with optional depth control."""

    path: str = Field(description="Directory path to list")
    depth: int | None = Field(
        default=None, description="Maximum child depth to include"
    )

    @field_validator("path")
    @classmethod
    def path_must_not_be_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Path cannot be blank")
        return v

    @field_validator("depth")
    @classmethod
    def depth_must_be_non_negative(cls, v: int | None) -> int | None:
        if v is not None and v < 0:
            raise ValueError("Depth must be non-negative")
        return v


class MoveEntry(BaseModel):
    """
    Request to move/rename a file or directory.

    Moves a file or directory from one location to another within the sandbox filesystem.
    Can be used for both renaming (same directory) and moving (different directory).
    """

    src: str = Field(
        description="Source path of the file or directory to move", alias="source"
    )
    dest: str = Field(
        description="Destination path where the file or directory should be moved",
        alias="destination",
    )

    @field_validator("src")
    @classmethod
    def src_must_not_be_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Source path cannot be blank")
        return v

    @field_validator("dest")
    @classmethod
    def dest_must_not_be_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Destination path cannot be blank")
        return v

    model_config = ConfigDict(populate_by_name=True)


class SetPermissionEntry(BaseModel):
    """
    Request to set permissions/ownership of a file or directory.

    Updates the permissions and/or ownership of an existing file or directory
    without modifying its content. Only specified properties will be changed.
    """

    path: str = Field(description="Target path of the file or directory to modify")
    owner: str | None = Field(default=None, description="New owner username")
    group: str | None = Field(default=None, description="New group name")
    mode: int = Field(default=755, description="New Unix file permissions as integer")

    @field_validator("path")
    @classmethod
    def path_must_not_be_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Path cannot be blank")
        return v

    @field_validator("mode")
    @classmethod
    def mode_must_be_non_negative(cls, v: int) -> int:
        if v < 0:
            raise ValueError("Mode must be non-negative")
        return v


class ContentReplaceEntry(BaseModel):
    """
    Request to replace content within a file.

    Performs string replacement within a file by finding exact matches of the old content
    and replacing them with new content. Only affects string matches, preserving the rest.
    """

    path: str = Field(description="Target file path containing content to replace")
    old_content: str = Field(
        description="Exact string content to find and replace", alias="old_content"
    )
    new_content: str = Field(
        description="Replacement string content to substitute", alias="new_content"
    )

    @field_validator("path")
    @classmethod
    def path_must_not_be_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Path cannot be blank")
        return v

    model_config = ConfigDict(populate_by_name=True)


class ContentReplaceResult(BaseModel):
    """
    Result of a content replacement operation on a single file.
    """

    path: str = Field(description="File path where replacement was performed")
    replaced_count: int = Field(description="Number of occurrences replaced. 0 means old content was not found.")


class SearchEntry(BaseModel):
    """
    Request to search for files matching a pattern.

    Searches the filesystem starting from the specified path to find files
    that match the given pattern. Used for file discovery and filtering.
    """

    path: str = Field(description="Starting directory path for the search")
    pattern: str = Field(
        description="Search pattern (supports glob patterns like *.py, *.txt)"
    )

    @field_validator("path")
    @classmethod
    def path_must_not_be_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Path cannot be blank")
        return v

    @field_validator("pattern")
    @classmethod
    def pattern_must_not_be_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Pattern cannot be blank")
        return v

File: sdks/sandbox/csharp/src/OpenSandbox/Models/Filesystem.cs
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Text.Json.Serialization;

namespace OpenSandbox.Models;

/// <summary>
/// Information about a file in the sandbox.
/// </summary>
public class SandboxFileInfo
{
    /// <summary>
    /// Gets or sets the file path.
    /// </summary>
    [JsonPropertyName("path")]
    public required string Path { get; set; }

    /// <summary>
    /// Gets or sets the entry type.
    /// </summary>
    [JsonPropertyName("type")]
    public string? Type { get; set; }

    /// <summary>
    /// Gets or sets the file size in bytes.
    /// </summary>
    [JsonPropertyName("size")]
    public long? Size { get; set; }

    /// <summary>
    /// Gets or sets the last modification time.
    /// </summary>
    [JsonPropertyName("modified_at")]
    public DateTime? ModifiedAt { get; set; }

    /// <summary>
    /// Gets or sets the creation time.
    /// </summary>
    [JsonPropertyName("created_at")]
    public DateTime? CreatedAt { get; set; }

    /// <summary>
    /// Gets or sets the file mode (permissions).
    /// </summary>
    [JsonPropertyName("mode")]
    public int? Mode { get; set; }

    /// <summary>
    /// Gets or sets the file owner.
    /// </summary>
    [JsonPropertyName("owner")]
    public string? Owner { get; set; }

    /// <summary>
    /// Gets or sets the file group.
    /// </summary>
    [JsonPropertyName("group")]
    public string? Group { get; set; }
}

/// <summary>
/// File permission settings.
/// </summary>
public class Permission
{
    /// <summary>
    /// Gets or sets the file mode (permissions).
    /// </summary>
    [JsonPropertyName("mode")]
    public int Mode { get; set; }

    /// <summary>
    /// Gets or sets the file owner.
    /// </summary>
    [JsonPropertyName("owner")]
    public string? Owner { get; set; }

    /// <summary>
    /// Gets or sets the file group.
    /// </summary>
    [JsonPropertyName("group")]
    public string? Group { get; set; }
}

/// <summary>
/// File metadata for upload operations.
/// </summary>
public class FileMetadata
{
    /// <summary>
    /// Gets or sets the file path.
    /// </summary>
    [JsonPropertyName("path")]
    public required string Path { get; set; }

    /// <summary>
    /// Gets or sets the file mode (permissions).
    /// </summary>
    [JsonPropertyName("mode")]
    public int? Mode { get; set; }

    /// <summary>
    /// Gets or sets the file owner.
    /// </summary>
    [JsonPropertyName("owner")]
    public string? Owner { get; set; }

    /// <summary>
    /// Gets or sets the file group.
    /// </summary>
    [JsonPropertyName("group")]
    public string? Group { get; set; }
}

/// <summary>
/// Entry for writing a file.
/// </summary>
public class WriteEntry
{
    /// <summary>
    /// Gets or sets the file path.
    /// </summary>
    public required string Path { get; set; }

    /// <summary>
    /// Gets or sets the file data.
    /// Supports: string, byte[], Stream.
    /// </summary>
    public object? Data { get; set; }

    /// <summary>
    /// Gets or sets the file mode (permissions).
    /// </summary>
    public int? Mode { get; set; }

    /// <summary>
    /// Gets or sets the file owner.
    /// </summary>
    public string? Owner { get; set; }

    /// <summary>
    /// Gets or sets the file group.
    /// </summary>
    public string? Group { get; set; }
}

/// <summary>
/// Entry for creating a directory.
/// </summary>
public class CreateDirectoryEntry
{
    /// <summary>
    /// Gets or sets the directory path.
    /// </summary>
    public required string Path { get; set; }

    /// <summary>
    /// Gets or sets the directory mode (permissions).
    /// </summary>
    public int? Mode { get; set; }

    /// <summary>
    /// Gets or sets the directory owner.
    /// </summary>
    public string? Owner { get; set; }

    /// <summary>
    /// Gets or sets the directory group.
    /// </summary>
    public string? Group { get; set; }
}

/// <summary>
/// Entry for searching files.
/// </summary>
public class SearchEntry
{
    /// <summary>
    /// Gets or sets the search path.
    /// </summary>
    public required string Path { get; set; }

    /// <summary>
    /// Gets or sets the search pattern (e.g., "*.txt").
    /// </summary>
    public string? Pattern { get; set; }
}

/// <summary>
/// Entry for moving/renaming a file.
/// </summary>
public class MoveEntry
{
    /// <summary>
    /// Gets or sets the source path.
    /// </summary>
    public required string Src { get; set; }

    /// <summary>
    /// Gets or sets the destination path.
    /// </summary>
    public required string Dest { get; set; }
}

/// <summary>
/// Entry for replacing content in a file.
/// </summary>
public class ContentReplaceEntry
{
    /// <summary>
    /// Gets or sets the file path.
    /// </summary>
    public required string Path { get; set; }

    /// <summary>
    /// Gets or sets the old content to replace.
    /// </summary>
    public required string OldContent { get; set; }

    /// <summary>
    /// Gets or sets the new content.
    /// </summary>
    public required string NewContent { get; set; }
}

/// <summary>
/// Entry for setting file permissions.
/// </summary>
public class SetPermissionEntry
{
    /// <summary>
    /// Gets or sets the file path.
    /// </summary>
    public required string Path { get; set; }

    /// <summary>
    /// Gets or sets the file mode (permissions).
    /// </summary>
    public required int Mode { get; set; }

    /// <summary>
    /// Gets or sets the file owner.
    /// </summary>
    public string? Owner { get; set; }

    /// <summary>
    /// Gets or sets the file group.
    /// </summary>
    public string? Group { get; set; }
}

/// <summary>
/// Options for reading a file as text.
/// </summary>
public class ReadFileOptions
{
    /// <summary>
    /// Gets or sets the text encoding (default: utf-8).
    /// </summary>
    public string? Encoding { get; set; }

    /// <summary>
    /// Gets or sets the byte range to read (e.g., "bytes=0-1023").
    /// Mutually exclusive with Offset/Limit.
    /// </summary>
    public string? Range { get; set; }

    /// <summary>
    /// Gets or sets the starting line number (1-based) for line-based reading.
    /// Mutually exclusive with Range.
    /// </summary>
    public int? Offset { get; set; }

    /// <summary>
    /// Gets or sets the number of lines to return for line-based reading.
    /// Mutually exclusive with Range.
    /// </summary>
    public int? Limit { get; set; }
}

/// <summary>
/// Options for reading a file as bytes.
/// </summary>
public class ReadBytesOptions
{
    /// <summary>
    /// Gets or sets the byte range to read (e.g., "bytes=0-1023").
    /// Mutually exclusive with Offset/Limit.
    /// </summary>
    public string? Range { get; set; }

    /// <summary>
    /// Gets or sets the starting line number (1-based) for line-based reading.
    /// Mutually exclusive with Range.
    /// </summary>
    public int? Offset { get; set; }

    /// <summary>
    /// Gets or sets the number of lines to return for line-based reading.
    /// Mutually exclusive with Range.
    /// </summary>
    public int? Limit { get; set; }
}

/// <summary>
/// API request model for renaming files.
/// </summary>
public class RenameFileItem
{
    /// <summary>
    /// Gets or sets the source path.
    /// </summary>
    [JsonPropertyName("src")]
    public required string Src { get; set; }

    /// <summary>
    /// Gets or sets the destination path.
    /// </summary>
    [JsonPropertyName("dest")]
    public required string Dest { get; set; }
}

/// <summary>
/// API request model for replacing file content.
/// </summary>
public class ReplaceFileContentItem
{
    /// <summary>
    /// Gets or sets the old content to replace.
    /// </summary>
    [JsonPropertyName("old")]
    public required string Old { get; set; }

    /// <summary>
    /// Gets or sets the new content.
    /// </summary>
    [JsonPropertyName("new")]
    public required string New { get; set; }
}

/// <summary>
/// Result of a content replacement operation on a single file.
/// </summary>
public class ContentReplaceResult
{
    /// <summary>
    /// Gets or sets the file path.
    /// </summary>
    public required string Path { get; set; }

    /// <summary>
    /// Gets or sets the number of occurrences replaced. 0 means old content was not found.
    /// </summary>
    public int ReplacedCount { get; set; }
}

/// <summary>
/// API response model for replace file content result.
/// </summary>
public class ReplaceFileContentResult
{
    /// <summary>
    /// Gets or sets the number of occurrences replaced.
    /// </summary>
    [JsonPropertyName("replacedCount")]
    public int ReplacedCount { get; set; }
}
